### ISSUE TITLE
Add like `editor.contentLeftPadding` to add padding between the `editorGutter` and `editor`.

### ISSUE THREAD

**Issue by shujaatak** [microsoft/vscode#135114]
Please add some option like `editor.contentLeftPadding` or whatever you want to name it, to resolve at least these two issues:
1. Currently in VS Code, the gutter and the code editor are so close that while clicking in the beginning of a line one can **accidentally click on collapse button** in the gutter region which sometimes really bother the user.

2. When a user apply a different background color to gutter for either theming purpose or to show that it's not part of the editor region, the **code in the editor** and also the **vertical lines** between the first pair of curly braces of code blocks basically **touch the gutter.** See the following first screenshot.

![image](https://user-images.githubusercontent.com/14161646/137374585-28ea3ca9-e545-4711-80ad-3811846c2a6d.png)

If we had  `editor.contentLeftPadding` then the user could just add that to settings.json like this:
```
"workbench.colorCustomizations": {
        "editor.contentLeftPadding": 8px;
        "editorGutter.background" : "#454545",
    },
```
which could separate code from the gutter and look like this:

![image](https://user-images.githubusercontent.com/14161646/137376181-999f40f6-d587-4fd2-839a-688472b5b919.png)

**Comment by alexdima** [comment]
cc @roblourens . I think you also needed something like this in Notebooks? (I noticed a commit from you)

**Comment by shujaatak** [comment]
@alexdima Can you please somehow expedite the process to resolve this issue? 
Almost all code editors provide **default feature to avoid the issues mentioned above** but unfortunately vs code lacks this.
**We all would be so grateful to you if you please resolve this issue as soon as possible.** By the way, I could have helped in resolving this but I don't know very well the internals of vs code.

**Comment by mckeed** [comment]
There are already settings for `editor.padding.top` and `editor.padding.bottom`, so the setting should be `editor.padding.left` and should result in the same amount of spacing as those do.

**Comment by realSilasYang** [comment]
When can we get the function satisfifed ... have noticed so many similar issues been proposed !!

**Comment by ishazel** [comment]
Hoping for a solution soon :)

**Comment by mckeed** [comment]
@pmeaney if you want less space in the gutter to the left of the line numbers, the ticket for that is https://github.com/microsoft/vscode/issues/93887 or https://github.com/microsoft/vscode/issues/30795 

Also, FYI, you can currently reduce the space to the right of the line numbers by setting `"editor.showFoldingControls": "never"`, and it looks like you have some other things enabled that are increasing the line decorations space (to the left of the line numbers) because mine is only 20px wide.

**Comment by shujaatak** [comment]
@alexdima Waiting for a year and counting :yawning_face:

**Comment by yCodeTech** [comment]
Just stumbled on this looking for this exact solution. Not sure why it's so difficult to get an answer, it's not like it's difficult to implement, as the preliminary work is already there for the padding top and bottom as previously said. Please can we get this implemented.

**Comment by ChristopherHaws** [comment]
I am constantly accidently hitting the collapse button when trying to select code starting at the start of a line! Hope this gets implemented at some point. :)

**Comment by mraheelkhan** [comment]
Maybe it helps someone. 

`"editor.lineDecorationsWidth" : 100`
`"editor.glyphMargin": false,`

It will put space between line start and line numbers/folding icon. 

I have plenty of space between the folding icon and line start. It helps me to avoid accidental clicks on folding.

**Comment by matprophet** [comment]
Would love to see this implemented.  "lineDecorationsWidth" gets my view part of the way there but a real margin would be a nice visual change.

**Comment by RichMorin** [comment]
I'm following up on @josevalim's [comment](https://github.com/livebook-dev/livebook/issues/2124#issuecomment-1657071864) in the Livebook thread just mentioned above. I'd like to suggest that there should be a way to make any additional padding clickable. This would make it easier to put the editing cursor (e.g., in Monaco) at the beginning of a line.

**Comment by realSilasYang** [comment]
There is merely no space where the arrow point in VS code …
![134806925-834b0200-12ce-44fe-90d1-93a9e9cab897](https://github.com/microsoft/vscode/assets/64590265/7625dabb-cc54-42cd-9e9f-ab2adeb6d63a)
While a little bit more space makes the whole look much better, as in PyCharm (and other IDEs)
![134808549-27c20288-2bc9-4698-b849-627dedb34dbb](https://github.com/microsoft/vscode/assets/64590265/505bfe30-90f9-48e2-8f5f-a433437f26a9)
Is it really that difficult to improve it? VSCode need to be more humble to learn from other famous editors.

**Comment by dschil138** [comment]
So is there really no solution for this? Is there an extension I could use or something? Even like 5 pixels of clickable text area to the left would make life so muck easier.

This seems like a relatively simple feature with a big QOL benefit, but this issue has been open for two years now. There are several duplicate issues that have been closed to reference back to this one. I see @alexdima was assigned at one point, but then maybe unassigned? I can't really tell.

Alex, do you know if this is being actively worked on at all, or did this slip through the cracks somehow?

**Comment by pmeaney** [comment]
his seems to take care of _some_ of the issue by allowing me to re-claim pixels from the number  bar.

**However, it would be great if I could collapse it even more, to even just 1-2 pixels of gutter on either side of the line numbers.**

settings.json

```json
{
    "editor.glyphMargin": false,
    "editor.folding": false,
    "editor.tabSize": 2,
}
```

<img width="116" alt="Screenshot 2023-10-03 at 6 15 25 PM" src="https://github.com/microsoft/vscode/assets/6299810/206ee470-403d-4239-82e0-814e5e580413">

**Comment by dschil138** [comment]
@pmeany I think part of the problem here is that we have two, almost contradictory issues bundled together. One issue is wanting to make the left gutter narrower, so that there is less wasted space, which you seem to have found a workable solution for.

The issue I'm concerned with is that I want an option to put *more* space to the left of the text, specifically space that I can still click with the text cursor, so that it is easier to select text to (or from) the beginning of a line.

IE: 5-10 pixels to the left of the start of each line, where if you click in those pixels, it puts the cursor at the beginning of that line (just to the right of where you clicked)

**Comment by RichMorin** [comment]
> On Oct 3, 2023, at 16:33, dschil138 ***@***.***> wrote:
> 
> ... The issue I'm concerned with is that I want an option to put more space to the left of the text, specifically space that I can still click with the text cursor, so that it is easier to select text to (or from) the beginning of a line.
> IE: 5-10 pixels to the left of the start of each line, where if you click in those pixels, it puts the cursor at the beginning of that line (just to the right of where you clicked)

Yep.  I'll even provide a WP quote...

> Fitts's law (often cited as Fitts' law) is a predictive model of human movement primarily used in human–computer interaction and ergonomics.  The law predicts that the time required to rapidly move to a target area is a function of the ratio between the distance to the target and the width of the target.  Fitts's law is used to model the act of pointing, either by physically touching an object with a hand or finger, or virtually, by pointing to an object on a computer monitor using a pointing device.  It was initially developed by Paul Fitts. ...

-- https://en.wikipedia.org/wiki/Fitts's_law

-r

**Comment by starball5** [comment]
Related on Stack Overflow:

- (canonical) [Add left space/margin/padding in VS Code editor window between line numbers and code](https://stackoverflow.com/q/59031656/11107541)
- (duplicate) [How can I add padding to VS Code's code editor panel](https://stackoverflow.com/q/63679736/11107541)
- (related) [Why do certain fonts' characters get cropped off the left margin of the editor panel in VS Code and how can it be fixed?](https://stackoverflow.com/q/75610946/11107541)

Recently on Reddit: https://www.reddit.com/r/vscode/comments/176br36/is_there_any_way_to_adjust_the_autospacing_at_the/

**Comment by dschil138** [comment]
@RichMorin thank you for the reply, I'm not sure I totally understand what you were saying though. I see that there were some GitHub actions triggered but I'm not super familiar with that system in relation to this project. Does this mean that this is being worked on? Or is coming soon?

**Comment by dschil138** [comment]
Yeah there seems to be a lot of interest in this issue. Lots of people trying to give answers in all those threads but none of them actually do the thing: Give more clickable space before the first character of the line.

**Comment by tkj-seasony** [comment]
I want to second this idea. Moving from CLion to VSCode, this difference is one of the things that frustrates me the most.
The missing editor padding from the gutter makes it less clear when reading unindented lines of code (i.e. function declarations etc.)

**Comment by mindThomas** [comment]
I want to second this idea. Moving from CLion to VSCode, this difference is one of the things that frustrates me the most.
The missing editor padding from the gutter makes it less clear when reading unindented lines of code (i.e. function declarations etc.), especially when the gutter color is not the same as the editor background.

**Comment by mocenigo** [comment]
I second this one. The missing left padding is a very annoying thing.

**Comment by mocenigo** [comment]
> @pmeany I think part of the problem here is that we have two, almost contradictory issues bundled together. One issue is wanting to make the left gutter narrower, so that there is less wasted space, which you seem to have found a workable solution for.
> 
> The issue I'm concerned with is that I want an option to put _more_ space to the left of the text, specifically space that I can still click with the text cursor, so that it is easier to select text to (or from) the beginning of a line.
> 
> IE: 5-10 pixels to the left of the start of each line, where if you click in those pixels, it puts the cursor at the beginning of that line (just to the right of where you clicked)

And if you start dragging, it will select whole lines.

**Comment by thoennes** [comment]
Still nothing?  This has been years, no?

**Comment by mocenigo** [comment]
It should not be difficult. The first change would be equivalent to adding
For proof of concept, I added the following CSS code to the end of this CSS file 

.monaco-editor .lines-content {
    padding-left: 8px;
}

(or some configurable value) to "vs/workbench/workbench.desktop.main.css" – and then to make sure that any rulers/guides are also drawn shifted by 8px (or the configured value). I am sure that for the devs it should be trivial.

**Comment by dschil138** [comment]
@mocenigo would that solution allow the space to the left of the first character to be clickable? That is crucial. In the first post by @shujaatak, he seemed to say that this approach would work on a visual level, but not allow the user to click in that space to set the cursor at the start of the line, I think.

Is this something that can be accomplished by a 3rd party in a pull request or something like that? I'm not super familiar with how development on this project works.

**Comment by mocenigo** [comment]
> @mocenigo would that solution allow the space to the left of the first character to be clickable? That is crucial. In the first post by @shujaatak, he seemed to say that this approach would work on a visual level, but not allow the user to click in that space to set the cursor at the start of the line, I think.

Yes, It should be clickable. Note that you can already select whole lines by clicking (and dragging) on the line numbers in the gutter.

**Comment by yCodeTech** [comment]
Why is this not even assigned to a dev? If I knew who to tag, I would. Devs, chime in please. This needs accelerating to the next version!!!!

**Comment by gjsjohnmurray** [comment]
@yCodeTech there are other issues with more upvotes.

Here is a query to see them

is:issue is:open sort:reactions-+1-desc milestone:Backlog

**Comment by pmeaney** [comment]
@yCodeTech 
@ @gjsjohnmurray Please reach out to OpenTeams (https://github.com/openteamsinc) to get their guidance on Companies which use VSCode to approach about this long-time-requested improvement?
If we can get OpenTeams to organize some corporate funding to be thrown at Devs on this, it would rise in priority rankings.
"a platform for open source maintainers and contributors to propose specific units of work that businesses can fund to accelerate open source technology roadmaps. This program helps to build a community of like-minded people who can work together to make open source projects better, ultimately leading to faster development and innovation." https://www.openteams.com/cooperative-funding/

@gjsjohnmurray I encourage you to submit a case study as "Requesting Funding to Prioritize a highly requested but lower-prioritized VS Code feature" at https://www.openteams.com/create-a-case-study-and-profile/
Thanks!

**Comment by dschil138** [comment]
@gjsjohnmurray I think part of the reason it has fewer upvotes is 1). The initial request was worded a little confusingly & also  kind of conflated two issues  2). the issue is over two years old, thus ranked very low chronologically.

There WAS a dev assigned to it, and it was even planned to be included for Oct 2021 updates, but then for some reason unassigned and never done.

I honestly think this is a situation where it was planned to be done, but then there was some kind of mixup and it fell through the cracks.

I don't claim to know how easy this is to accomplish but several others have mentioned that it should be a relatively simple feature to implement.

**Comment by leafstrat** [comment]
Little hack since this was bugging me. 
I'm using the [Todo Tree extension](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree), which gives me icons in the gutter, and I didn't like how they looked lacking "editor left padding".

Since **I don't use the minimap**, I thought maybe I could use it as a padding buffer, and it works perfectly.

[editor_left_padding.webm](https://github.com/microsoft/vscode/assets/4014775/766cd138-7327-4b6c-8ece-6c812a6113fd)

```js
{
  // "editor left padding"
  "editor.minimap.enabled": true,
  "editor.minimap.autohide": true,
  "editor.minimap.side": "left",
  "editor.minimap.maxColumn": 6,
  // ---
  // gutter settings
  "editor.lineDecorationsWidth": 6,
  "editor.lineNumbers": false,
  "editor.folding": false,
  // ---
  // gutter styling
  "workbench.colorCustomizations": {
  "editorGutter.background": "#555"
  }, // ---
}
```

**Comment by dschil138** [comment]
@leafstrat can you click in there and have the cursor go to the beginning of that line?

**Comment by mocenigo** [comment]
I just saw the various "workaround" and I see that some make wrong assumpitons. 
The padding on the left should be at the beginning of the rendered line. So if the first visible column is the first one, you have some spacing between the gutter and the text. If, however, the text is scrolled a bit, then you will see a fraction of text rendered in that padding space.

However if we add

.monaco-editor .lines-content {
padding-left: 8px;
}

to "vs/workbench/workbench.desktop.main.css" 

then the vertical rulers defined by  "editor.rulers" in settings.json will not be aligned with the text. In other words, these rulers should be also shifted by 8px. However, this does not happen with the .monaco-editor .lines-content  hack.

 Roberto

**Comment by dschil138** [comment]
@mocenigo sorry could you rephrase that, I'm not sure I follow. Are you saying that you found a solution?

**Comment by mocenigo** [comment]
> @mocenigo sorry could you rephrase that, I'm not sure I follow. Are you saying that you found a solution?
> 
> **EDIT:** just tried adding the style you included in the comment. It definitely solves the visual element. Unfortunately it appears to be Drake #2 from my meme, unclickable padding :/
> 
> I may be misunderstanding if that specific style was the solution that you were suggesting tho

Hi no, I did not find a solution, and my apologies for not answering earlier. That only adds  some inactive space and doe not work well with vertical rulers.

I just discovered that there is an option called editor.letterSpacing and this adds spacing between characters in pixels and the vertical rulers work well. So, a parameter like editor.spacingBeforeFirstLetter if implemented in the same way would probably work – in other words the changes could be made to the letter spacing code instead of the buffer rendering code.

**Comment by dschil138** [comment]
I created a new issue in order to try and revive this old issue. If you still care about this please go give it your support!

**Comment by pmeaney** [comment]
Hi @dschil138 Thanks for actively pursing this improvement!

Here's what I've gone with---

**- To reclaim horizontal space:**

 these settings:  https://github.com/microsoft/vscode/issues/30795#issuecomment-410998882
"editor.glyphMargin": false,
"editor.folding": false, // this will disable the vertical column which allows for minimizing code sections i.e. “folding”
"editor.lineNumbers": "off",
"workbench.activityBar.visible": false



**- To reclaim vertical space:**

Source 1: https://github.com/microsoft/vscode/issues/174903#issuecomment-1726455954
Source 2: Similar / related: https://github.com/drcika/apc-extension/issues/52#issuecomment-1675837688

1. install extension [Apc Customize UI++](https://marketplace.visualstudio.com/items?itemName=drcika.apc-extension)
2. add these lines to your user settings.json:

"window.titleBarStyle": "native",
"apc.electron": {
   "titleBarStyle": "customButtonsOnHover",
},
"apc.header": {
   "height": 20
},
"apc.sidebar.titlebar": {
   "height": 20
},
"apc.activityBar": {
   "size": 77,
   "itemSize": 48,
   "itemMargin": 0
},
"apc.stylesheet": {
   ".custom-sidebar-titlebar .sidebar .composite.title": "padding-left: 0;", // Don't indent the sidebar title. We want it horizontally aligned to the left so that it makes a nice vertical column with the controls in the sidebar.
   ".custom-sidebar-titlebar .sidebar .composite.title .inline-titlebar-placeholder": "padding-left: 0 !important;", // Don't indent the sidebar title. Part 2 to remove the empty space to the left of the sidebar title and complete the proper horizontal left alignment.
   ".monaco-workbench .part.statusbar>.items-container>.statusbar-item.left.first-visible-item": "padding-left: 0;", // Don't indent the statusbar items. This horizontally aligns the statusbar items with the "Problems" header in the panel, which looks nice.
},
// END -- settings to hide top menu bar and other stylesheet polish

## Example of Vertical space efficiency improvements:

<img width="1095" alt="Screenshot 2024-03-07 at 5 53 05 PM" src="https://github.com/microsoft/vscode/assets/6299810/05513e9e-ac34-4e05-bdb5-996de11f747d">

**Comment by RichMorin** [comment]
Thanks for keeping this on the radar. I'll freely admit that I don't understand all of the CSS you posted, but I'm happy that you're trying...

**Comment by dschil138** [comment]
y'all gotta go comment on the new issue or this is never getting done. This is our last stand!

**Comment by Adam-Kay** [comment]
> ```js
> "editorGutter.background": "#555"
> ```

Exact same issue with TODO, but I'd prefer to keep the minimap. Any ideas on how to achieve it differently?

**Comment by MariuzM** [comment]
Damn 3 years passed still no `editor.padding.left` eh

**Comment by erdemdev** [comment]
Devs do something

**Comment by deflock** [comment]
> Devs do something

Who needs those boring paddings when AI is now hype train. Choo-choo!🚂

**Comment by ajtompkins** [comment]
Just moving from notepad++ to vscode and how hard should it be to select the first character of a line? Not this hard.

**Comment by pmeaney** [comment]
> Just moving from notepad++ to vscode and how hard should it be to select the first character of a line? Not this hard.

The issue has shifted to this one: https://github.com/microsoft/vscode/issues/206914
thanks to @dschil138 .  @ajtompkins if you comment on that issue I link above, we may get better traction there.

**Comment by cooltools2077** [comment]
@yCodeTech  I don't plan to publish the plugin yet. If you'd like to use it, you can install it directly from the source code.

1. Package the Extension: Use the following command in your terminal: npx vsce package

2. Install the VSIX File in VS Code: In VS Code, go to the Extensions view, click the More Actions (...) menu, and select Install from VSIX... Choose the .vsix file you just built.
